### PR TITLE
[4.0] Redo of custom elements tabs

### PR DESCRIPTION
**Pull request by dgrammatiko**
Pull Request for many Issues originating from the tabs script

fixes:
https://github.com/joomla/joomla-cms/issues/33229
https://github.com/joomla/joomla-cms/issues/34492
https://github.com/joomla/joomla-cms/issues/31583

### Summary of Changes
- The logic was rewritten from scratch
- Same funtionality
- Better code
- Support for dynamically added tabs
- A11y fixes (based on the WAI-ARIA recommendations)

Changes of the source js/css can be found: https://github.com/joomla-projects/custom-elements/pull/180

### Testing Instructions
Test each and every form in the backend that has tabs

### Expected result AFTER applying this Pull Request

Tabs View:
Screenshot 2021-07-17 at 14 06 02

Screenshot 2021-07-17 at 14 06 14

Shrink to accordion:
Screenshot 2021-07-17 at 14 07 03

Screenshot 2021-07-17 at 14 07 03

### Documentation Changes Required

- Added the recall and breakpoint attributes to the expected values for the creation of the tabs (UiTabs::)
- The `section` element (the tab content) is renamed to `joomla-tab-element`

### CODE REVIEW

File: libraries/src/HTML/Helpers/UiTab.php [hunk 1]
```diff
@@ -79,7 +82,7 @@ public static function endTabSet()
 	 *
 	 * @param   string  $selector  Identifier of the panel.
 	 * @param   string  $id        The ID of the div element
-	 * @param   string  $title     The title text for the new UL tab
+	 * @param   string  $title     The title text for the button
```
Comment (brianteeman) [inline libraries/src/HTML/Helpers/UiTab.php:86]: button ??

File: plugins/content/pagebreak/pagebreak.php [hunk 2]
```diff
@@ -221,7 +221,7 @@ public function onContentPrepare($context, &$row, &$params, $page = 0)
 
 				if ($style === 'tabs')
 				{
-					$t[] = (string) HTMLHelper::_('uitab.startTabSet', 'myTab', array('active' => 'article' . $row->id . '-' . $style . '0'));
+					$t[] = (string) HTMLHelper::_('uitab.startTabSet', 'myTab', array('active' => 'article' . $row->id . '-' . $style .'0', 'recall' => true, 'breakpoint' => 768));
```
Comment (brianteeman) [inline plugins/content/pagebreak/pagebreak.php]:
```suggestion
					$t[] = (string) HTMLHelper::_('uitab.startTabSet', 'myTab', array('active' => 'article' . $row->id . '-' . $style . '0', 'recall' => true, 'breakpoint' => 768));
```

File: libraries/src/HTML/Helpers/UiTab.php (hunk 1 above)
Comment (dgrammatiko) [reply libraries/src/HTML/Helpers/UiTab.php:86]: It's the tab/accordion button that switches the tab/accordion (the navigation part)

File: libraries/src/HTML/Helpers/UiTab.php [hunk 3]
```diff
@@ -54,12 +54,15 @@ public static function startTabSet($selector = 'myTab', $params = array())
...
 		}
 
-		// @TODO echo the recall attribute correctly, now it's hardcoded!!!
-		$recall = !isset($params['recall']) ? '' : 'recall';
-
 		$orientation = isset($params['orientation']) ? $params['orientation'] : 'horizontal';
+		$recall = isset($params['recall']) ? 'recall' : '';
+		$breakpoint  = isset($params['breakpoint']) ? 'breakpoint="' . $params['breakpoint'] . '"' : '';
+
+		if (!isset($params['breakpoint']) && $breakpoint === '') {
```
Comment (richard67) [inline libraries/src/HTML/Helpers/UiTab.php]:
```suggestion
		if (!isset($params['breakpoint']) && $breakpoint === '')
		{
```
Fix PHP CS error in Drone.

File: libraries/src/HTML/Helpers/UiTab.php [hunk 4]
```diff
@@ -54,12 +54,15 @@ public static function startTabSet($selector = 'myTab', $params = array())
...
+			$breakpoint = 'breakpoint="768"';
+		}
 
-		return '<joomla-tab id="' . $selector . '" orientation="' . $orientation . '" recall>';
+		return '<joomla-tab id="' . $selector . '" orientation="' . $orientation . '" ' . $recall .' ' . $breakpoint . '>';
```
Comment (richard67) [inline libraries/src/HTML/Helpers/UiTab.php]:
```suggestion
		return '<joomla-tab id="' . $selector . '" orientation="' . $orientation . '" ' . $recall . ' ' . $breakpoint . '>';
```
Fix PHP CS error in Drone.

**Comment by richard67** [comment]
@dgrammatiko Now PHPCS is fine, but npm is failing in Drone. Could that be the same problem as with your other PR?

https://ci.joomla.org/joomla/joomla-cms/45867/1/1

```
npm ci --unsafe-perm
npm ERR! code ENOENT
npm ERR! syscall spawn git
npm ERR! path git
npm ERR! errno -2
npm ERR! enoent Error while executing:
npm ERR! enoent undefined ls-remote -h -t https://github.com/joomla-projects/custom-elements.git
npm ERR! enoent 
npm ERR! enoent 
npm ERR! enoent spawn git ENOENT
npm ERR! enoent This is related to npm not being able to find a file.
```

**Comment by dgrammatiko** [comment]
> Could that be the same problem as with your other PR?

Yes, it's exactly the same issue: the node alpine container doesn't have the git installed.
I guess this limits the testers to only the ones that can git pull, etc (pretty sure also the created packages are broken)

**Comment by richard67** [comment]
> I guess this limits the testers to only the ones that can git pull, etc (pretty sure also the created packages are broken)

@dgrammatiko Correct.

File: administrator/components/com_banners/tmpl/banner/edit.php [hunk 5]
```diff
@@ -29,7 +29,7 @@
 	<?php echo LayoutHelper::render('joomla.edit.title_alias', $this); ?>
 
 	<div class="main-card">
-		<?php echo HTMLHelper::_('uitab.startTabSet', 'myTab', array('active' => 'details')); ?>
+		<?php echo HTMLHelper::_('uitab.startTabSet', 'myTab', array('active' =>'details', 'recall' => true, 'breakpoint' => 768)); ?>
```
Comment (Quy) [inline administrator/components/com_banners/tmpl/banner/edit.php]:
```suggestion
		<?php echo HTMLHelper::_('uitab.startTabSet', 'myTab', array('active' => 'details', 'recall' => true, 'breakpoint' => 768)); ?>
```

Comment (Quy) [reply administrator/components/com_banners/tmpl/banner/edit.php]: Please fix the same with other instances.

**Comment by Quy** [comment]
This displays quickly then disappears.

joomla-stats

**Comment by Quy** [comment]
Text should be left-aligned.

permissions

**Comment by dgrammatiko** [comment]
> This displays quickly then disappears.

did you test the https://github.com/joomla/joomla-cms/pull/34781 before? Sounds like this is comming from there

**Comment by Quy** [comment]
This is a clean install from the download page of this PR.

**Comment by dgrammatiko** [comment]
> This is a clean install from the download page of this PR.

Ahh, forgot that George already merged a PR for alerts that wasn't quite ready, treat anything alerts related here as irrelevant

**Comment by Quy** [comment]
Permissions is selected.

global-configuration-permissions

Go to a component. No tab is selected by default.

global-configuration

**Comment by dgrammatiko** [comment]
@Quy should ok now
Screenshot 2021-07-17 at 20 09 02

**Comment by Quy** [comment]
Go to Global Configuration.
Click on one of the components.
Styling issue on the first tab.

tab-style

**Comment by Quy** [comment]
In mobile view, text should be left-aligned. The vertical ellipse is not visible against the blue background.

tab-mobile

**Comment by Quy** [comment]
Go to System > Administrator Templates > Atum Details and Files

tab-editor

tab-create-overrides

**Comment by Quy** [comment]
Edit an article.
Hover over a tab.
The label becomes bold causing words/tabs to the right of this tab to shift.

**Comment by Quy** [comment]
Will PR #31419 be obsolete because of this PR?

**Comment by dgrammatiko** [comment]
> The label becomes bold causing words/tabs to the right of this tab to shift.

Fixed

**Comment by dgrammatiko** [comment]
> Will PR #31419 be obsolete because of this PR?

With the provided code (modified to use more attributes):
```html
<?= HTMLHelper::_('uitab.startTabset', 'tabs1', ['active' => 'tabA', 'view' => 'tabs', 'orientation' => 'horizontal']) ?>
<?= HTMLHelper::_('uitab.addTab', 'tabs1', 'tabs11', 'tabs11') ?>
Horizontal tab 1

<?= HTMLHelper::_('uitab.startTabset', 'tabs2', ['active' => 'tabs21', 'view' => 'tabs', 'orientation' => 'vertical']) ?>
<?= HTMLHelper::_('uitab.addTab', 'tabs2', 'tabs21', 'tabs21') ?>
Horizontal &gt; Vertical tab 1

<?= HTMLHelper::_('uitab.startTabset', 'tabs3', ['active' => 'tabs31', 'view' => 'tabs', 'orientation' => 'horizontal']) ?>
<?= HTMLHelper::_('uitab.addTab', 'tabs3', 'tabs31', 'tabs31') ?>
Horizontal &gt; Vertical &gt; Horizontal tab 1
<?= HTMLHelper::_('uitab.endTab') ?>
<?= HTMLHelper::_('uitab.addTab', 'tabs3', 'tabs32', 'tabs32') ?>
Horizontal &gt; Vertical &gt; Horizontal tab 2
<?= HTMLHelper::_('uitab.endTab') ?>
<?= HTMLHelper::_('uitab.endTabset') ?>

<?= HTMLHelper::_('uitab.endTab') ?>

<?= HTMLHelper::_('uitab.addTab', 'tabs2', 'tabs22', 'tabs22') ?>
Horizontal &gt; Vertical tab 2
<?= HTMLHelper::_('uitab.endTab') ?>

<?= HTMLHelper::_('uitab.endTabset') ?>

<?= HTMLHelper::_('uitab.endTab') ?>

<?= HTMLHelper::_('uitab.addTab', 'tabs1', 'tabs12', 'tabs12') ?>
Horizontal tab 2
<?= HTMLHelper::_('uitab.endTab') ?>

<?= HTMLHelper::_('uitab.endTabset') ?>

```

I get: 
Screenshot 2021-07-17 at 23 12 14

(which is the expected output)

**Comment by brianteeman** [comment]
testing the changing of permissions for an article. After each change there is the following javascript error twice

```
joomla-tab.min.js?0.0.44:1 
Uncaught TypeError: t.getAttribute is not a function
    at HTMLElement.removeNavs (joomla-tab.min.js?0.0.44:1)
    at joomla-tab.min.js?0.0.44:1
    at Array.map (<anonymous>)
    at HTMLElement.onMutation (joomla-tab.min.js?0.0.44:1)
```

**Comment by brianteeman** [comment]
I have tested this item :red_circle: unsuccessfully on 4f0407cf129312ed884b1dd8be1568bb6c146b14<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34813">issues.joomla.org/tracker/joomla-cms/34813</a>.</sub>

**Comment by dgrammatiko** [comment]
@brianteeman was it only these console errors or there are more bugs?

btw I already pushed some fixes

**Comment by brianteeman** [comment]
Yes - I stopped at that point to do other stuff. retesting now

**Comment by brianteeman** [comment]
js issue is fixed.

there is however a css issue

before this PR
image


with this PR
image

**Comment by dgrammatiko** [comment]
> there is however a css issue

@brianteeman should be ok now

**Comment by brianteeman** [comment]
confirm that the tree is fixed now.

Spotted something different not which I think is something to do with focus-visible
tabs

**Comment by dgrammatiko** [comment]
> Spotted something different not which I think is something to do with focus-visible

The focus should work as expected now

**Comment by brianteeman** [comment]
> The focus should work as expected now

Still needs fixing here 
image

**Comment by Quy** [comment]
No background color on the active tab. Padding is reduced.

BEFORE PR:
34813-tinymce-before

AFTER PR:
34813-tinymce-after

**Comment by dgrammatiko** [comment]
> Still needs fixing here

That was another logic, should be ok now, also fixes another bug with image editing

**Comment by dgrammatiko** [comment]
@Quy done

**Comment by Quy** [comment]
34821-permissions

**Comment by Quy** [comment]
When editing/closing an image.

34821-undefined

34821-media-js

**Comment by Quy** [comment]
Edit an article. See padding less with PR.

BEFORE:
34821-content-pre
34821-images-pre

AFTER:
34821-content-after
34821-images-after

**Comment by dgrammatiko** [comment]
> When editing/closing an image.

Fixed.

Reviewing all the padding/margin values

**Comment by Quy** [comment]
Go to Media.
Edit an image.
Click Resize tab.
It is random when the image won't display.
Click Close button.
Edit the same image.
Image displays find under all tabs.

34813-media

**Comment by dgrammatiko** [comment]
A note here for testers: 

## The image editing is broken here but there's a new PR for that: https://github.com/joomla/joomla-cms/pull/34885

**Comment by Quy** [comment]
Click User Menu > Accessibility Settings
The `Accessibility Settings` tab is not selected.

File: administrator/components/com_installer/tmpl/install/default.php [hunk 6]
```diff
@@ -51,7 +51,7 @@
 					<?php endif; ?>
 
 					<?php if ($tabs) : ?>
-						<?php echo HTMLHelper::_('uitab.startTabSet', 'myTab', ['active' => $tabs[0]['name'] ?? '']); ?>
+						<?php echo HTMLHelper::_('uitab.startTabSet', 'myTab', ['active' => $tabs[0]['name'] ??'', 'recall' => true, 'breakpoint' => 768]); ?>
```
Comment (Quy) [inline administrator/components/com_installer/tmpl/install/default.php]:
```suggestion
						<?php echo HTMLHelper::_('uitab.startTabSet', 'myTab', ['active' => $tabs[0]['name'] ?? '', 'recall' => true, 'breakpoint' => 768]); ?>
```

**Comment by dgrammatiko** [comment]
> The Accessibility Settings tab is not selected.

Good catch,  fixed with https://github.com/joomla/joomla-cms/pull/34813/commits/bae7dfe2b807cbcf4f7175da5b60eea71547568c

**Comment by Quy** [comment]
Click User Menu > Accessibility Settings
Accessibility Settings tab is selected.
Click Close button.
Click User Menu > Accessibility Settings
Tab outlined. Click anywhere on the screen to refresh/fix the outline.

Firefox:
34813-firefox

Chrome:
34813-chrome

**Comment by brianteeman** [comment]
didnt i report that before?

**Comment by dgrammatiko** [comment]
> didnt i report that before?

You did and I fixed it, but I added the hash activation part that I forgot and I forgot to pass the `false` param.
Anyways it's fixed with https://github.com/joomla/joomla-cms/pull/34813/commits/cbce7b1113f4802543589a1eef8a29134bb7c281

File: administrator/components/com_plugins/tmpl/plugin/edit.php [hunk 7]
```diff
@@ -32,11 +32,11 @@
 <form action="<?php echo Route::_('index.php?option=com_plugins&view=plugin&layout=' . $layout . $tmpl . '&extension_id=' . (int) $this->item->extension_id); ?>" method="post" name="adminForm" id="style-form" aria-label="<?php echo Text::_('COM_PLUGINS_FORM_EDIT'); ?>" class="form-validate">
 	<div class="main-card">
 
-		<?php echo HTMLHelper::_('uitab.startTabSet', 'myTab', array('active' => 'general')); ?>
+		<?php echo HTMLHelper::_('uitab.startTabSet', 'myTab', ['active' => 'general', 'recall' => true, 'breakpoint' => 768]); ?>
 
 		<?php echo HTMLHelper::_('uitab.addTab', 'myTab', 'general', Text::_('COM_PLUGINS_PLUGIN')); ?>
 
-		<div class="row mt-2">
+		<div class="row mt-2 ps-4">
```
Comment (Quy) [inline administrator/components/com_plugins/tmpl/plugin/edit.php]:
```suggestion
		<div class="row mt-2">
```

34813-plugin

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 8f79a32a776e250e34408a9d12cd276102bca483<br /><br />THANK YOU!!!!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34813">issues.joomla.org/tracker/joomla-cms/34813</a>.</sub>

**Comment by dgrammatiko** [comment]
@Quy thank you for all the tests here  🙌

**Comment by ceford** [comment]
I have tested this item :white_check_mark: successfully on 8f79a32a776e250e34408a9d12cd276102bca483<br /><br />Went through most of the edit form and did not find any problems.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34813">issues.joomla.org/tracker/joomla-cms/34813</a>.</sub>

**Comment by richard67** [comment]
Maintainers please note that this PR should not be merged as it is. It needs to merge the corresponding upstream PR and make a new tag or release upstream and then change here to let npm use that. It will not need new tests after that if the new upstream tag or release point to the same upstream commit as used by this PR here now.

**Comment by dgrammatiko** [comment]
There's one more step, so, here's the whole procedure:
- on the Custom Elements Repo the https://github.com/joomla-projects/custom-elements/pull/180 needs to be merged
- A new patch version (0.2.0) needs to be released
- This PR needs to update the package.json and package.lock
- The PR #34885 needs to merge immediately after (assuming that what I did there is acceptable)

But since this PR is using the commit hash no further tests will be needed if the PR https://github.com/joomla-projects/custom-elements/pull/180 is merged without any changes, that was the reason I used the commit hashes here in the first place. Anyways these are @wilsonge's problems  😀

**Comment by ceford** [comment]
Something I neglected to check: the Article edit form has 12 html validation errors.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34813">issues.joomla.org/tracker/joomla-cms/34813</a>.</sub>

**Comment by dgrammatiko** [comment]
> Something I neglected to check: the Article edit form has 12 html validation errors.

Can you post the report here?

**Comment by ceford** [comment]
> > Something I neglected to check: the Article edit form has 12 html validation errors.
> 
> Can you post the report here?

It is probably best to paste the page source into the w3C validator:

https://validator.w3.org/#validate_by_input

30 Errors on the article edit page. It is not easy to copy them into a message.

**Comment by dgrammatiko** [comment]
@ceford you're validating as HTML5, right?
Screenshot 2021-07-24 at 14 28 10

**Comment by ceford** [comment]
Yes - I use the Firefox Web Developer plugin. It is just two clicks to invoke the w3c validator with default settings. I get 30 Errors for the Article edit page with no patches applied. It returns a blank page if I have Joomla debug enabled.

**Comment by dgrammatiko** [comment]
> I get 30 Errors for the Article edit page with no patches applied

Yes, I could see also the errors but none of them is about tabs

>  It returns a blank page if I have Joomla debug enabled.

Probably because the page is extremely heavy with debug on (too many elements)

**Comment by wilsonge** [comment]
@dgrammatiko i've released the custom elements repo. If you could use the tagged version here then one last test by someone just to do final validation would be excellent.

File: administrator/components/com_banners/tmpl/banner/edit.php [hunk 8]
```diff
@@ -29,7 +29,7 @@
 	<?php echo LayoutHelper::render('joomla.edit.title_alias', $this); ?>
 
 	<div class="main-card">
-		<?php echo HTMLHelper::_('uitab.startTabSet', 'myTab', array('active' => 'details')); ?>
+		<?php echo HTMLHelper::_('uitab.startTabSet', 'myTab', array('active' => 'details', 'recall' => true, 'breakpoint' => 768)); ?>
```
Comment (Quy) [inline administrator/components/com_banners/tmpl/banner/edit.php]:
```suggestion
		<?php echo HTMLHelper::_('uitab.startTabSet', 'myTab', ['active' => 'details', 'recall' => true, 'breakpoint' => 768]); ?>
```

File: plugins/content/pagebreak/pagebreak.php [hunk 9]
```diff
@@ -221,7 +221,7 @@ public function onContentPrepare($context, &$row, &$params, $page = 0)
 
 				if ($style === 'tabs')
 				{
-					$t[] = (string) HTMLHelper::_('uitab.startTabSet', 'myTab', array('active' => 'article' . $row->id . '-' . $style . '0'));
+					$t[] = (string) HTMLHelper::_('uitab.startTabSet', 'myTab', array('active' => 'article' . $row->id . '-' . $style . '0', 'view' => 'tabs'));
```
Comment (Quy) [inline plugins/content/pagebreak/pagebreak.php]:
```suggestion
					$t[] = (string) HTMLHelper::_('uitab.startTabSet', 'myTab', ['active' => 'article' . $row->id . '-' . $style . '0', 'view' => 'tabs']);
```

**Comment by dgrammatiko** [comment]
Tests should be ok with https://github.com/joomla-projects/joomla-browser/pull/219 (run fine loacally)

**Comment by wilsonge** [comment]
OK Can we have one final test of the code here now we're pulling in the stable custom elements repo and I'll get everything merged through.

**Comment by richard67** [comment]
> 
> 
> OK Can we have one final test of the code here now we're pulling in the stable custom elements repo and I'll get everything merged through.

@wilsonge And what about failing system tests?

**Comment by wilsonge** [comment]
The PR above needs to be merged according to @dgrammatiko - that would make the 4.0-dev fail if i merged that first - so i'll merge both at the same time and should be fine. I can debug further if anything else still fails.

**Comment by richard67** [comment]
It's late here already and I'm tired. Maybe @Quy can test again?

**Comment by Quy** [comment]
Border thickness is thinner.

Before PR:
34813-before

After PR (thickness difference between first row and second row):
34813-after

34813-after-2nd-row

**Comment by dgrammatiko** [comment]
@Quy pixel restored

**Comment by Quy** [comment]
There is a large gap at the bottom of the tab.

Before PR:
34813-global-before

After PR:
34813-global-after

**Comment by Quy** [comment]
Fieldset left/right borders missing.

Before PR:
34813-messages-before

After PR:
34813-messages-after

**Comment by dgrammatiko** [comment]
@Quy should be ok now

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on c75abd12b564df4114d89620da7ddc64ca7cddb8<br /><br />THANK YOU AGAIN!!!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34813">issues.joomla.org/tracker/joomla-cms/34813</a>.</sub>

**Comment by Quy** [comment]
I don't see errors in the Console.

System > Install > Extensions > Install from Web

34813-install

**Comment by richard67** [comment]
SCSS style linter complains: https://ci.joomla.org/joomla/joomla-cms/46132/1/20

**Comment by dgrammatiko** [comment]
> System > Install > Extensions > Install from Web

Web Installer fixed

> SCSS style linter complains: https://ci.joomla.org/joomla/joomla-cms/46132/1/20

CS fixed

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on c75abd12b564df4114d89620da7ddc64ca7cddb8<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/34813">issues.joomla.org/tracker/joomla-cms/34813</a>.</sub>

**Comment by wilsonge** [comment]
Thanks!

**Comment by dgrammatiko** [comment]
Thanks!

**Comment by dgrammatiko** [comment]
@wilsonge you need to merge the https://github.com/joomla-projects/joomla-browser/pull/219 and also the https://github.com/joomla/joomla-cms/pull/34885

The  https://github.com/joomla-projects/joomla-browser/pull/219 will fix the failing tests

**Pull request merged**